### ISSUE TITLE
Compilation Issue - Ubuntu 16.04

### ISSUE THREAD

**Issue by brettallred** [SpiderLabs/ModSecurity#1685]
**Steps to Reproduce** 
1. Spin up a fresh Ubuntu 16.04 instance on Digital Ocean (or some other provider)
2. Update system 
```
apt update
apt dist-upgrade
```
3. Install Pre Requisite dependencies
```
sudo apt install -y apt-utils autoconf automake build-essential git libcurl4-openssl-dev / 
libgeoip-dev liblmdb-dev libpcre++-dev libtool libxml2-dev libyajl-dev pkgconf /
wget zlib1g-dev
```

4. Download and Build ModSecurity V3
```
cd /usr/local/src
git clone --depth 1 -b v3/master --single-branch https://github.com/SpiderLabs/ModSecurity
cd ModSecurity
git submodule init
git submodule update
./build.sh
./configure
$ make
```

5. Error

```
/bin/bash ../../libtool  --tag=CXX   --mode=link g++  -g -O2  -lpcre -L/usr/lib/x86_64-linux-gnu/    -lpcre -o modsec-rules-check modsec_rules_check-rules-check.o -L/usr/lib/x86_64-linux-gnu -lcurl -lGeoIP   -lrt   -lxml2 -llmdb  -lpcre -lpcre  ../../src/.libs/libmodsecurity.a -lyajl
libtool: link: g++ -g -O2 -o modsec-rules-check modsec_rules_check-rules-check.o  -L/usr/lib/x86_64-linux-gnu/ -L/usr/lib/x86_64-linux-gnu /usr/lib/x86_64-linux-gnu/libcurl.so -lGeoIP -lrt -lxml2 -llmdb -lpcre ../../src/.libs/libmodsecurity.a -lyajl -pthread
../../src/.libs/libmodsecurity.a(libmodsecurity_la-validate_dtd.o): In function `modsecurity::operators::ValidateDTD::init(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/usr/local/src/ModSecurity/src/operators/validate_dtd.cc:37: undefined reference to `xmlThrDefSetGenericErrorFunc'
/usr/local/src/ModSecurity/src/operators/validate_dtd.cc:40: undefined reference to `xmlSetGenericErrorFunc'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-validate_dtd.o): In function `modsecurity::operators::ValidateDTD::evaluate(modsecurity::Transaction*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/operators/validate_dtd.cc:49: undefined reference to `xmlParseDTD'
/usr/local/src/ModSecurity/src/operators/validate_dtd.cc:85: undefined reference to `xmlNewValidCtxt'
/usr/local/src/ModSecurity/src/operators/validate_dtd.cc:98: undefined reference to `xmlValidateDtd'
/usr/local/src/ModSecurity/src/operators/validate_dtd.cc:111: undefined reference to `xmlFreeValidCtxt'
/usr/local/src/ModSecurity/src/operators/validate_dtd.cc:102: undefined reference to `xmlFreeValidCtxt'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-validate_dtd.o): In function `modsecurity::operators::ValidateDTD::~ValidateDTD()':
/usr/local/src/ModSecurity/src/../src/operators/validate_dtd.h:44: undefined reference to `xmlFreeDtd'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-validate_dtd.o): In function `modsecurity::operators::ValidateDTD::~ValidateDTD()':
/usr/local/src/ModSecurity/src/../src/operators/validate_dtd.h:44: undefined reference to `xmlFreeDtd'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-validate_schema.o): In function `modsecurity::operators::ValidateSchema::evaluate(modsecurity::Transaction*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:46: undefined reference to `xmlSchemaNewParserCtxt'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:63: undefined reference to `xmlSchemaSetParserErrors'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:66: undefined reference to `xmlThrDefSetGenericErrorFunc'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:69: undefined reference to `xmlSetGenericErrorFunc'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:71: undefined reference to `xmlSchemaParse'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:87: undefined reference to `xmlSchemaNewValidCtxt'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:102: undefined reference to `xmlSchemaSetValidErrors'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:131: undefined reference to `xmlSchemaValidateDoc'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:136: undefined reference to `xmlSchemaFree'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:137: undefined reference to `xmlSchemaFreeParserCtxt'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:83: undefined reference to `xmlSchemaFreeParserCtxt'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:145: undefined reference to `xmlSchemaFree'
/usr/local/src/ModSecurity/src/operators/validate_schema.cc:146: undefined reference to `xmlSchemaFreeParserCtxt'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-validate_schema.o): In function `modsecurity::operators::ValidateSchema::~ValidateSchema()':
/usr/local/src/ModSecurity/src/../src/operators/validate_schema.h:56: undefined reference to `xmlSchemaFreeValidCtxt'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-validate_schema.o): In function `modsecurity::operators::ValidateSchema::~ValidateSchema()':
/usr/local/src/ModSecurity/src/../src/operators/validate_schema.h:56: undefined reference to `xmlSchemaFreeValidCtxt'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-verify_cc.o): In function `modsecurity::operators::VerifyCC::init(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/usr/local/src/ModSecurity/src/operators/verify_cc.cc:96: undefined reference to `pcre_compile'
/usr/local/src/ModSecurity/src/operators/verify_cc.cc:103: undefined reference to `pcre_study'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-verify_cc.o): In function `modsecurity::operators::VerifyCC::~VerifyCC()':
/usr/local/src/ModSecurity/src/operators/verify_cc.cc:37: undefined reference to `pcre_free'
/usr/local/src/ModSecurity/src/operators/verify_cc.cc:42: undefined reference to `pcre_free_study'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-verify_cc.o): In function `modsecurity::operators::VerifyCC::evaluate(modsecurity::Transaction*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/operators/verify_cc.cc:129: undefined reference to `pcre_exec'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-https_client.o): In function `modsecurity::Utils::HttpsClient::download(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/utils/https_client.cc:69: undefined reference to `curl_easy_init'
/usr/local/src/ModSecurity/src/utils/https_client.cc:76: undefined reference to `curl_easy_setopt'
/usr/local/src/ModSecurity/src/utils/https_client.cc:78: undefined reference to `curl_slist_append'
/usr/local/src/ModSecurity/src/utils/https_client.cc:79: undefined reference to `curl_slist_append'
/usr/local/src/ModSecurity/src/utils/https_client.cc:91: undefined reference to `curl_easy_setopt'
/usr/local/src/ModSecurity/src/utils/https_client.cc:94: undefined reference to `curl_easy_setopt'
/usr/local/src/ModSecurity/src/utils/https_client.cc:95: undefined reference to `curl_easy_setopt'
/usr/local/src/ModSecurity/src/utils/https_client.cc:98: undefined reference to `curl_easy_setopt'
/usr/local/src/ModSecurity/src/utils/https_client.cc:101: undefined reference to `curl_easy_setopt'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-https_client.o):/usr/local/src/ModSecurity/src/utils/https_client.cc:103: more undefined references to `curl_easy_setopt' follow
../../src/.libs/libmodsecurity.a(libmodsecurity_la-https_client.o): In function `modsecurity::Utils::HttpsClient::download(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/utils/https_client.cc:113: undefined reference to `curl_easy_perform'
/usr/local/src/ModSecurity/src/utils/https_client.cc:115: undefined reference to `curl_slist_free_all'
/usr/local/src/ModSecurity/src/utils/https_client.cc:121: undefined reference to `curl_easy_cleanup'
/usr/local/src/ModSecurity/src/utils/https_client.cc:118: undefined reference to `curl_easy_strerror'
/usr/local/src/ModSecurity/src/utils/https_client.cc:110: undefined reference to `curl_easy_setopt'
/usr/local/src/ModSecurity/src/utils/https_client.cc:87: undefined reference to `curl_slist_append'
/usr/local/src/ModSecurity/src/utils/https_client.cc:83: undefined reference to `curl_slist_append'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-regex.o): In function `modsecurity::Utils::Regex::Regex(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/utils/regex.cc:51: undefined reference to `pcre_compile'
/usr/local/src/ModSecurity/src/utils/regex.cc:54: undefined reference to `pcre_study'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-regex.o): In function `modsecurity::Utils::Regex::~Regex()':
/usr/local/src/ModSecurity/src/utils/regex.cc:60: undefined reference to `pcre_free'
/usr/local/src/ModSecurity/src/utils/regex.cc:65: undefined reference to `pcre_free_study'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-regex.o): In function `modsecurity::Utils::Regex::searchAll(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/utils/regex.cc:83: undefined reference to `pcre_exec'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-regex.o): In function `modsecurity::Utils::regex_search(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, modsecurity::Utils::SMatch*, modsecurity::Utils::Regex const&)':
/usr/local/src/ModSecurity/src/utils/regex.cc:113: undefined reference to `pcre_exec'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-regex.o): In function `modsecurity::Utils::regex_search(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, modsecurity::Utils::Regex const&)':
/usr/local/src/ModSecurity/src/utils/regex.cc:127: undefined reference to `pcre_exec'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-modsecurity.o): In function `modsecurity::ModSecurity::ModSecurity()':
/usr/local/src/ModSecurity/src/modsecurity.cc:82: undefined reference to `curl_global_init'
/usr/local/src/ModSecurity/src/modsecurity.cc:85: undefined reference to `xmlInitParser'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-modsecurity.o): In function `modsecurity::ModSecurity::~ModSecurity()':
/usr/local/src/ModSecurity/src/modsecurity.cc:92: undefined reference to `curl_global_cleanup'
/usr/local/src/ModSecurity/src/modsecurity.cc:98: undefined reference to `xmlCleanupParser'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-xml.o): In function `modsecurity::RequestBodyProcessor::XML::~XML()':
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:38: undefined reference to `xmlFreeParserCtxt'
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:42: undefined reference to `xmlFreeDoc'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-xml.o): In function `modsecurity::RequestBodyProcessor::XML::init()':
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:56: undefined reference to `xmlParserInputBufferCreateFilenameDefault'
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:53: undefined reference to `__xmlParserInputBufferCreateFilename'
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:53: undefined reference to `xmlParserInputBufferCreateFilenameDefault'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-xml.o): In function `modsecurity::RequestBodyProcessor::XML::processChunk(char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:109: undefined reference to `xmlParseChunk'
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:95: undefined reference to `xmlCreatePushParserCtxt'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-xml.o): In function `modsecurity::RequestBodyProcessor::XML::complete(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:126: undefined reference to `xmlParseChunk'
/usr/local/src/ModSecurity/src/request_body_processor/xml.cc:133: undefined reference to `xmlFreeParserCtxt'
../../src/.libs/libmodsecurity.a(libmodsecurity_la-geo_lookup.o): In function `modsecurity::operators::GeoLookup::evaluate(modsecurity::Transaction*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/operators/geo_lookup.cc:66: undefined reference to `GeoIP_region_name_by_code'
/usr/local/src/ModSecurity/src/operators/geo_lookup.cc:94: undefined reference to `GeoIPRecord_delete'
../../src/.libs/libmodsecurity.a(lt2-libmodsecurity_la-geo_lookup.o): In function `modsecurity::Utils::GeoLookup::cleanUp()':
/usr/local/src/ModSecurity/src/utils/geo_lookup.cc:45: undefined reference to `GeoIP_delete'
../../src/.libs/libmodsecurity.a(lt2-libmodsecurity_la-geo_lookup.o): In function `modsecurity::Utils::GeoLookup::setDataBase(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
/usr/local/src/ModSecurity/src/utils/geo_lookup.cc:51: undefined reference to `GeoIP_open'
../../src/.libs/libmodsecurity.a(lt2-libmodsecurity_la-geo_lookup.o): In function `modsecurity::Utils::GeoLookup::lookup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, GeoIPRecordTag**, std::function<bool (int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)>)':
/usr/local/src/ModSecurity/src/utils/geo_lookup.cc:70: undefined reference to `GeoIP_record_by_name'
../../src/.libs/libmodsecurity.a(lt5-libmodsecurity_la-xml.o): In function `modsecurity::Variables::XML::evaluate(modsecurity::Transaction*, modsecurity::Rule*, std::vector<modsecurity::VariableValue const*, std::allocator<modsecurity::VariableValue const*> >*)':
/usr/local/src/ModSecurity/src/variables/xml.cc:83: undefined reference to `xmlXPathNewContext'
/usr/local/src/ModSecurity/src/variables/xml.cc:99: undefined reference to `xmlXPathRegisterNs'
/usr/local/src/ModSecurity/src/variables/xml.cc:127: undefined reference to `xmlXPathFreeObject'
/usr/local/src/ModSecurity/src/variables/xml.cc:146: undefined reference to `xmlXPathFreeContext'
/usr/local/src/ModSecurity/src/variables/xml.cc:116: undefined reference to `xmlXPathEvalExpression'
/usr/local/src/ModSecurity/src/variables/xml.cc:142: undefined reference to `xmlFree'
/usr/local/src/ModSecurity/src/variables/xml.cc:135: undefined reference to `xmlNodeGetContent'
collect2: error: ld returned 1 exit status
Makefile:540: recipe for target 'modsec-rules-check' failed
make[2]: *** [modsec-rules-check] Error 1
make[2]: Leaving directory '/usr/local/src/ModSecurity/tools/rules-check'
Makefile:477: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/usr/local/src/ModSecurity/tools'
Makefile:954: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

This appears to be an issue referencing libxml2 but that dependency is installed.

**Comment by petarkozic** [comment]
Same problem !

**Comment by zimmerle** [comment]
Hi @brettallred and @petarkozic,

Working in the fix at this very moment.

**Comment by zimmerle** [comment]
Fixed by: 854a661a2a86357326b5c6be9799e9ab943b61e3
